pkg/controller/apiextensions/requirement/configurator.go

```diff
@@ -0,0 +1,66 @@
+/*
+Copyright 2019 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package requirement
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/crossplane/crossplane-runtime/apis/core/v1alpha1"
+	"github.com/crossplane/crossplane-runtime/pkg/fieldpath"
+	"github.com/crossplane/crossplane-runtime/pkg/meta"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+	"github.com/crossplane/crossplane-runtime/pkg/resource/unstructured/composite"
+	"github.com/crossplane/crossplane-runtime/pkg/resource/unstructured/requirement"
+)
+
+// Configure the supplied composite resource. The composite resource name is
+// derived from the supplied requirement, as {namespace}-{name}-{random-string}.
+// The requirement's external name annotation, if any, is propagated to the
+// composite resource.
+func Configure(_ context.Context, rq resource.Requirement, cp resource.Composite) error {
+	cp.SetGenerateName(fmt.Sprintf("%s-%s-", rq.GetNamespace(), rq.GetName()))
+	if meta.GetExternalName(rq) != "" {
+		meta.SetExternalName(cp, meta.GetExternalName(rq))
+	}
+
+	urq, ok := rq.(*requirement.Unstructured)
+	if !ok {
+		return nil
+	}
+	ucp, ok := cp.(*composite.Unstructured)
+	if !ok {
+		return nil
+	}
+
+	// YOLO. This seems to work; it copies over some unknown fields which are
+	// dropped and not stored due to the CRD being configured to do so.
+	spec, _ := fieldpath.Pave(urq.Object).GetValue("spec")
+	_ = fieldpath.Pave(ucp.Object).SetValue("spec", spec)
```

Comment:
I think this should be fixed, too, before merging even though it looks safe I don't feel comfortable with full copy. Should be trivial do `delete()` the keys from the map since we already know the keys that shouldn't make it to composite.

Reply:
Honestly I don't love either approach. I worry about the maintainability of having to update a blacklist of fields that should be deleted from. The alternative would be to require that all user specified fields were nested in (for example) `spec.parameters` or require that all Crossplane specific fields were nested under something like `spec.requirement`. I think this might be worth considering in future because it clearly distinguishes user specified fields from injected, Crossplane specific fields.